File: entities/projectile.py
# entities/projectile.py
import math
import pygame
import random
from settings import *


class Projectile:
    def __init__(self, x, y, angle):
        self.x = x
        self.y = y
        self.angle = angle
        self.speed = 15
        self.radius = 4
        self.life_time = 100
        self.marked_for_deletion = False
        self.anim_offset = random.uniform(0, 10)

    def update(self, game_map):
        self.life_time -= 1
        if self.life_time <= 0:
            self.marked_for_deletion = True
            return

        dx = math.cos(self.angle) * self.speed
        dy = math.sin(self.angle) * self.speed

        if game_map.get_wall(int((self.x + dx) // TILE_SIZE), int((self.y + dy) // TILE_SIZE)):
            self.marked_for_deletion = True
        else:
            self.x += dx
            self.y += dy

    def draw_2d(self, surface, offset_x, offset_y, zoom):
        sx = self.x * zoom + offset_x
        sy = self.y * zoom + offset_y

        pulse = math.sin(pygame.time.get_ticks() * 0.05 + self.anim_offset) * 2
        size = int((self.radius + pulse) * zoom)
        size = max(2, size)

        color_core = (200, 255, 0)
        color_glow = (100, 200, 0, 100)

        tail_len = 20 * zoom
        ex = sx - math.cos(self.angle) * tail_len
        ey = sy - math.sin(self.angle) * tail_len
        pygame.draw.line(surface, color_glow, (sx, sy), (ex, ey), int(2 * zoom))

        pygame.draw.circle(surface, color_core, (sx, sy), size)

File: src/entities/projectile.rs
// src/entities/projectile.rs
use crate::config::*;
use crate::map::MapHandler;

pub struct Projectile {
    pub x: f32,
    pub y: f32,
    pub angle: f32,
    pub speed: f32,
    pub radius: f32,
    pub life_time: i32,
    pub marked_for_deletion: bool,
}

impl Projectile {
    pub fn new(x: f32, y: f32, angle: f32) -> Self {
        Self {
            x, y, angle,
            speed: 15.0,
            radius: 4.0,
            life_time: 100,
            marked_for_deletion: false,
        }
    }

    pub fn update(&mut self, map: &mut MapHandler) {
        self.life_time -= 1;
        if self.life_time <= 0 {
            self.marked_for_deletion = true;
            return;
        }

        let dx = self.angle.cos() * self.speed;
        let dy = self.angle.sin() * self.speed;

        if map.get_wall(
            ((self.x + dx) / TILE_SIZE) as i32,
            ((self.y + dy) / TILE_SIZE) as i32,
        ) != 0 {
            self.marked_for_deletion = true;
        } else {
            self.x += dx;
            self.y += dy;
        }
    }
}
